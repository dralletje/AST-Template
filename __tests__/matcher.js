let { template, minivaluate } = require('../app.js');

it('should minivaluate some expression', () => {
  expect(minivaluate(template.expression`'Hey'`.ast)).toEqual('Hey');
  expect(minivaluate(template.expression`80 + 90`.ast)).toEqual(80 + 90);
  expect(minivaluate(template.expression`'He' + 'y'`.ast)).toEqual('Hey');
  expect(minivaluate(template.expression`{ id: 10 }`.ast)).toEqual({ id: 10 });
});

it('should match a simple sequelize definition', () => {
  let scope = {
    DataTypes: {
      UUID: 'UUID',
      UUIDV4: 'UUIDV4',
      STRING: 'STRING',
      DECIMAL: 'DECIMAL',
    },
  };
  let simple_definition = `
    module.exports = (sequelize, DataTypes) => {
      const Price = sequelize.define(
        'Price',
        {
          id: {
            type: DataTypes.UUID,
            primaryKey: true,
            defaultValue: DataTypes.UUIDV4,
            allowNull: false,
            autoIncrement: false,
          },
          shape: DataTypes.STRING,
          color: DataTypes.STRING,
          clarity: DataTypes.STRING,
          lowSize: DataTypes.DECIMAL,
          highSize: DataTypes.DECIMAL,
          caratPrice: DataTypes.DECIMAL, // TODO Use integers (javascript decimals are broken (0.1 + 0.2 !== 0.3))
          date: DataTypes.STRING,
        },
      );

      return Price;
    };
  `;

  let definition_template = template.statements`
    module.exports = (sequelize, DataTypes) => {
      const ${template.Identifier('model')} = sequelize.define(
        ${template.String('model_name')},
        ${template.Object('model_definition', {
          key: template.expression`${template.Identifier('property')}`,
          value: template.expression`${template.Expression('type_description', {
            scope: scope,
          })}`,
        })}
      );

      return ${template.Identifier('model')};
    };
  `;

  expect(definition_template.match(simple_definition)).toMatchSnapshot();
});

it.only('should match a definition with optionals', () => {
  let simple_definition = `
    let result = { id: 'hey' };
    // result.name = 'Cool';
    module.exports = result;
  `;

  let definition_template = template.statements`
    let ${template.Identifier('var')} = ${template.Object('value', {
      key: template.expression`${template.Identifier('property')}`,
      value: template.expression`${template.String('type_description')}`,
    })};

    ${template.repeat('optional_found', 0, 1, template.expression`${template.Expression('type_description')}`)}

    module.exports = ${template.Identifier('var')};
  `;

  expect(definition_template.match(simple_definition)).toMatchSnapshot();
});
