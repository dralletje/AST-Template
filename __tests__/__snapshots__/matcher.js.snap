// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match a definition with optionals 1`] = `
Object {
  "areas": Object {
    "value": Object {
      "mismatches": Array [],
      "value": Array [
        Object {
          "key": Object {
            "property": "id",
          },
          "value": Object {
            "type_description": "hey",
          },
        },
      ],
    },
    "var": "result",
  },
  "mismatches": Array [],
}
`;

exports[`should match a simple sequelize definition 1`] = `
Object {
  "areas": Object {
    "model": "Price",
    "model_definition": Object {
      "mismatches": Array [],
      "value": Array [
        Object {
          "key": Object {
            "property": "id",
          },
          "value": Object {
            "type_description": Object {
              "allowNull": false,
              "autoIncrement": false,
              "defaultValue": "UUIDV4",
              "primaryKey": true,
              "type": "UUID",
            },
          },
        },
        Object {
          "key": Object {
            "property": "shape",
          },
          "value": Object {
            "type_description": "STRING",
          },
        },
        Object {
          "key": Object {
            "property": "color",
          },
          "value": Object {
            "type_description": "STRING",
          },
        },
        Object {
          "key": Object {
            "property": "clarity",
          },
          "value": Object {
            "type_description": "STRING",
          },
        },
        Object {
          "key": Object {
            "property": "lowSize",
          },
          "value": Object {
            "type_description": "DECIMAL",
          },
        },
        Object {
          "key": Object {
            "property": "highSize",
          },
          "value": Object {
            "type_description": "DECIMAL",
          },
        },
        Object {
          "key": Object {
            "property": "caratPrice",
          },
          "value": Object {
            "type_description": "DECIMAL",
          },
        },
        Object {
          "key": Object {
            "property": "date",
          },
          "value": Object {
            "type_description": "STRING",
          },
        },
      ],
    },
    "model_name": "Price",
  },
  "mismatches": Array [],
}
`;
